### ISSUE TITLE
[HOLD for payment 2022-12-20] Investigate simple code splitting in webpack bundle

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#12235]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open a new incognito window
1. Open the developer console, network tab
1. Throttle your network connection to Slow 3G
1. Go to NewDot web (production or staging)

## Expected Result:
The site should load and become visible reasonably quickly, even over a slow network.

## Actual Result:
The site takes about 55 seconds to load.

## Workaround:
n/a

## Platform:
Where is this issue occurring?

- Web
- Mobile Web

**Issue reported by:** @roryabraham 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1666947230935069

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~019fbb45bb5ac42437</li>
        <li>Upwork Job ID: 1606464380411809792</li>
        <li>Last Price Increase: 2022-12-24</li>
    </ul>
</details>

**Comment by roryabraham** [comment]
Created this issue specifically for @kidroca to investigate implementing the simple "no downside" improvements described in the slack thread. Not as interested in investing in more lazy loading than that right now. Let's see how much we can improve this without having to influence the main code development.

**Comment by kidroca** [comment]
Sure thing
Configuring vendor chunks would also aid webpack development speed as caching would be more effective

**Comment by kidroca** [comment]
Just to be clear:

The improvements with vendor chunks, would probably not help much for new users (incognito mode) loading the web app 
It's intended to help existing users refreshing/reopening after an update has been deployed

The improvement in load time we might see for this (vendor chunks) change is - fresh loading time might improve a bit from fetching smaller chunks in parallel

**Comment by roryabraham** [comment]
Assuming no update here, which is fine. I think @kidroca has a few other higher priority issues

**Comment by roryabraham** [comment]
Same as last week ... @kidroca when do you think you can get a PR up to split out vendor code into separate chunk(s)? Seems like it might be reasonably quick and would be good to have this going in parallel

**Comment by kidroca** [comment]
Yeah, I've been putting this off, because it's not a priority

If it's not a problem that someone would have to review/test it, I can squeeze a PR next week

**Comment by kidroca** [comment]
### Current load times
- simulated slow 3g
- incognito

<details>
<summary>No cache - 60sec.</summary>

<img width="945" alt="Screenshot 2022-11-26 at 14 16 16" src="https://user-images.githubusercontent.com/12156624/204088536-b519c500-0ee0-4660-8152-569ac7d47b3a.png">
</details>

<details>
<summary>With Browser cache - 6.53 sec.</summary>

<img width="945" alt="Screenshot 2022-11-26 at 14 20 38" src="https://user-images.githubusercontent.com/12156624/204088680-22931ffb-f310-458e-b2a1-cbe15b9d8197.png">
</details>

**Comment by roryabraham** [comment]
Yeah, that is not great. This isn't a high priority so only work on it where you have spare cycles for now but keep me posted and I can review a PR.

**Comment by kidroca** [comment]
### With bundle splitting
- one `main.js` file of 1.78 MB
- one big `vendors.js` file of 11.28 MB

Initial load times didn't change. It still takes about 60 sec. to load App on a "slow 3g", but partial caching is now possible - meaning every time we're only making changes to App code, users would have to download only the smaller `main.js` file 
`vendors.js` would change when we make package.json dependencies changes - updating/adding/removing external libraries including `expnesify-common` or `react-native-onyx`

--- 

There are other strategies like configuring Webpack to extract a js file for each 3rd party library, but this would only improve partial caching, while load times stay the same.

Webpack maintainers recommend to stay with the built in strategy that creates one `vendors.js` chunk and instead focus on code splitting (using `async import`)
- this would create more chunks naturally from modularized App code
- it promotes a modular and more maintainable App structure

Only code splitting would improve initial load speed for uncached loads 

---

One more thing worth mentioning - the backend seems to be configured to expire `.js` files in 4hrs 
When those 4hrs expire we'll be downloading the whole app again, even if it didn't change

When using Webpack we don't need to configure the server to expire `.js` files, webpack appends unique hashes based on the script content. We can just serve the `.js` files and browsers would cache them, when files change they would be served with a different hash and invalidate browser cache

**Comment by kidroca** [comment]
### With code splitting (just one chunk) (2x faster initial load) 
- one `main.js` file of 1.78 MB
- one big `vendors.js` file of 6.1 MB
- one big `onfido-sdk.js` file of 5.11 MB
- "slow 3g" first time loading - 30.8s, loading the following in parallel:
    - 10.76s for `main.js` 
    - 24.33s for `vendors.js`
    - n/a onfido - not loaded on the initial page

### Onfido SDK is almost 1/2 of the total App size
<details>
<summary>Before code splitting</summary>

![image](https://user-images.githubusercontent.com/12156624/205014120-23bb8ed4-015e-499b-8dc3-f213ece3cd4f.png)
</details>

<details>
<summary>With code splitting</summary>

![image](https://user-images.githubusercontent.com/12156624/205015003-054c83d4-6878-4821-98bc-ddb4d526a3ce.png)

_This 5.11 MB code chunk is now only loaded when we're at the Onfido step_
</details>

### POC for code splitting

#### Simple code splitting using async code and dynamic import

https://github.com/Expensify/App/blob/22e0a24177f7a2909866d06d53026c394ab52f68/src/components/Onfido/BaseOnfidoWeb.js#L20-L26
_Component mounts and runs async code_
_Optionally the component can render something until the `import()` promise resolves_

#### Code splitting using `React.lazy` 

##### Use lazy import 

https://github.com/Expensify/App/blob/429554d3e66e5e403780d404b1ce2910382f1320/src/components/Onfido/index.website.js#L6-L9

_Replace a regular import statement with a `React.lazy` one liner_
_(Almost) everything about this lazy component would be captured in a separate bundle and loaded on demand_

##### Use `<Suspense>` when rendering the lazy component

https://github.com/Expensify/App/blob/429554d3e66e5e403780d404b1ce2910382f1320/src/components/Onfido/index.website.js#L26-L37

_When the lazy component have not been or is still loading a fallback content can be displayed_

### Preload demo

This video show App loading in 30 sec on a slow network and only then continues to load the remaining (Onfido) content 
_(App remains usable in the meantime)_

https://user-images.githubusercontent.com/12156624/205067384-2496d7d2-45f6-4b3d-9a0c-51c5d00fb6e8.mp4

**Comment by roryabraham** [comment]
> the backend seems to be configured to expire .js files in 4hrs

How did you discover that? Maybe I can look into this piece

**Comment by roryabraham** [comment]
> Onfido SDK is almost 1/2 of the total App size 

This is good to know, since we're getting rid of it at some point soon. [Internal issue link here](https://github.com/Expensify/Expensify/issues/233363)

**Comment by roryabraham** [comment]
[This POC](https://github.com/Expensify/App/pull/13232) looks good to me, but given that we'll be getting rid of the Onfido SDK soon, I'd say let's keep it as just a POC and only move forward with https://github.com/Expensify/App/pull/13235, because that doesn't really establish a new code pattern that we need broader consensus on

**Comment by kidroca** [comment]
> > the backend seems to be configured to expire .js files in 4hrs
> 
> How did you discover that? Maybe I can look into this piece

Just load/reload `new.expensify.com` with devtools opened on the Network tab
Optionally enable filtration to only see JS requests
Find the `main.js` script and inspect response headers:

```
cache-control: public, max-age=14400
expires: Sat, 03 Dec 2022 00:54:15 GMT
                // now is 21:05:00 GMT
```

<img width="789" alt="Screenshot 2022-12-02 at 22 58 20" src="https://user-images.githubusercontent.com/12156624/205385540-a03108b0-f771-4032-8d22-dc8e9cc2a662.png">

Seems to come from Cloudflare dashboard setting

**Comment by bondydaa** [comment]
I've asked someone to help us look into cloudflare's settings to see where that `max-age` is being set b/c I don't see it anywhere in our other config settings.

**Comment by bondydaa** [comment]
hmm though the hash on the main.js bundle seems to always be the same `db426a36985fca224c47` - this is currently the hash I get on production and staging and the same thing from your screenshot (at least from what I can see)

prod:
<img width="1512" alt="image (12)" src="https://user-images.githubusercontent.com/4073354/205692380-cd1ded13-450c-40c0-95cf-ca6079436edf.png">


staging:
<img width="1512" alt="image (11)" src="https://user-images.githubusercontent.com/4073354/205692397-2e9baaa5-0d8e-4c17-8e2f-c4c774b46c3e.png">

**Comment by kidroca** [comment]
> hmm though the hash on the main.js bundle seems to always be the same db426a36985fca224c47 - this is currently the hash I get on production and staging and the same thing from your screenshot

Interesting, perhaps a new version haven't been released yet 
I'll inveterate why staging and prod have the same hash - hash should be based on the content and even though these envs are pretty much the same there are slight differences 

Even if the hash for staging and prod is the same as long as it changes when new build is published it should be OK

**Comment by bondydaa** [comment]
Yeah that's what I was trying to verify as well, at the very least with local dev when I changed something it generated a new hash 
<img width="1512" alt="image" src="https://user-images.githubusercontent.com/4073354/205704131-165e4eb7-a9d5-4219-9436-9284011a1424.png">

we should have a new deploy today right that gets pushed to staging so we should be able to verify once that's done that the deploy process is properly changing hashes as well?

**Comment by kidroca** [comment]
@bondydaa 
Hashes are different for staging and prod, in your comment here: https://github.com/Expensify/App/issues/12235#issuecomment-1337692

You see the file from `new.expensify.com` - not the one from `staging.new.expensify.com` 
<img width="272" alt="Screenshot 2022-12-05 at 20 23 41" src="https://user-images.githubusercontent.com/12156624/205713190-797e59c3-952f-4e1f-80e6-49b8a44aa3f7.png">
556 

I guess you first opened `new.` then `staging.new.`, this would keep the request in the network history, so when you search for `db426a36985fca224c47` - it's the only thing you see 
You should instead look for `main-`

**Comment by bondydaa** [comment]
oh lol doh 🤦 good catch

**Comment by bondydaa** [comment]
discussing how we want to handle the CF cache setting, I think we can probably just up it to like 30 days or a year (a year is what we have on old dot since we do something similar with hashes for the JS files). 

That way we get the benefit of caching but then anytime a new hash is deployed the user will have to redownload it.

**Comment by roryabraham** [comment]
> discussing how we want to handle the CF cache setting

@bondydaa Should we create a separate issue to handle this in a follow-up?

**Comment by mallenexpensify** [comment]
@aimane-chnaif can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~019fbb45bb5ac42437

**Comment by mallenexpensify** [comment]
Paid @aimane-chnaif .  This doesn't look like it needs regression test steps added, comment/reopen if anyone disagrees

**Comment by bondydaa** [comment]
we upped the cloudflare cache TTL to 1 day for now on https://github.com/Expensify/Expensify/issues/251424. 

we can always tweak it more if needed.
